File: leetcode/215.数组中的第K个最大元素/215.数组中的第K个最大元素_1.py
"""
二叉堆解法-小顶堆
把小顶堆 small 理解成一个筛子，较大的元素会沉淀下去，较小的元素会浮上来；

当堆大小超过 k 的时候，删掉堆顶的元素，因为这些元素比较小。
遍历一遍后，小顶堆里面留下第k个大的元素就是堆顶，堆底是最大的
"""
import heapq
from typing import List


class Solution:
    def findKthLargest(self, nums: List[int], k: int) -> int:
        small = []
        for num in nums:
            heapq.heappush(small, num)
            if len(small) > k:
                heapq.heappop(small)
        return small[0]


if __name__ == '__main__':
    nums = [3, 2, 1, 5, 6, 4]
    k = 2
    result = Solution().findKthLargest(nums, k)
    print(result)

File: leetcode/215.数组中的第K个最大元素/215.数组中的第K个最大元素_1.go
package _15_数组中的第K个最大元素

import (
	"container/heap"
	"sort"
)

// 二叉堆解法-小顶堆
type hp struct {
	sort.IntSlice
}

func (h *hp) Push(v interface{}) {
	h.IntSlice = append(h.IntSlice, v.(int))
}
func (h *hp) Pop() interface{} {
	a := h.IntSlice
	v := a[len(a)-1]
	h.IntSlice = a[:len(a)-1]
	return v
}

func findKthLargest(nums []int, k int) int {
	small := &hp{}
	for _, num := range nums {
		heap.Push(small, num)
		if small.Len() > k {
			heap.Pop(small)
		}
	}
	return small.IntSlice[0]
}
